### PR TITLE
Improve exception safety of mutex usage

### PR DESCRIPTION
**Pull request by wkaluza**
There is still some code in DKG, network and the HttpConnectionManager which is unsafe, but it doesn't lend itself to an immediate fix.

### CODE REVIEW

**Review by ejfitzgerald — CHANGES_REQUESTED** [review]
Looks good, but can we switch to the `FETCH_LOCK` macro please

File: libs/network/include/network/details/idle_work_store.hpp [hunk 1]
```diff
@@ -92,7 +92,13 @@ class IdleWorkStore
...
     if (mutex_.try_lock())
     {
-      is_due = Clock::now() >= (last_run_ + interval_);
+      try
+      {
+        is_due = Clock::now() >= (last_run_ + interval_);
+      }
+      catch (...)
+      {
+      }
```
Comment (ejfitzgerald) [inline libs/network/include/network/details/idle_work_store.hpp:101]: Should we log this case do you think?

**Review by ejfitzgerald — DISMISSED** [review]

**Review by ejfitzgerald — APPROVED** [review]

File: libs/network/include/network/details/idle_work_store.hpp (hunk 1 above)
Comment (wkaluza) [reply libs/network/include/network/details/idle_work_store.hpp:101]: In summary of our conversation on Slack just now, `idle_work_store.hpp` is scheduled for demolition in the near future.

File: libs/core/src/mutex.cpp [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+namespace {
+constexpr std::size_t DEFAULT_TIMEOUT_MS = 3000;
+}
+
+/**
+ * The debug mutex acts like a normal mutex but also contains several other checks. This code is
+ * intended to be only used in software development.
+ */
+DebugMutex::MutexTimeout::MutexTimeout(std::string filename, int const &line)
```
Comment (bipll) [inline libs/core/src/mutex.cpp:48]: `int const &`

File: libs/core/src/mutex.cpp [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+  LOG_STACK_TRACE_POINT;
+  FETCH_LOG_ERROR(LOGGING_NAME, "The system will terminate, mutex timed out: ", filename_, " ",
+                  line_);
+
+  // Send a sigint to ourselves since we have a handler for this
+  kill(0, SIGINT);
+}
+
+DebugMutex::DebugMutex(int line, std::string file)
+  : AbstractMutex()
```
Comment (bipll) [inline libs/core/src/mutex.cpp:96]: Is this explicit default-init needed?

**Pull request merged**